libevmasm/Assembly.cpp

```diff
@@ -1796,47 +1801,46 @@ LinkerObject const& Assembly::assembleEOF() const
 	return ret;
 }
 
-std::vector<size_t> Assembly::decodeSubPath(size_t _subObjectId) const
+std::vector<SubAssemblyID> Assembly::decodeSubPath(SubAssemblyID _subObjectId) const
 {
-	if (_subObjectId < m_subs.size())
+	if (_subObjectId.value < m_subs.size())
 		return {_subObjectId};
 
-	auto subIdPathIt = find_if(
-		m_subPaths.begin(),
-		m_subPaths.end(),
+	auto subIdPathIt = ranges::find_if(
+		m_subPaths,
 		[_subObjectId](auto const& subId) { return subId.second == _subObjectId; }
 	);
 
 	assertThrow(subIdPathIt != m_subPaths.end(), AssemblyException, "");
 	return subIdPathIt->first;
 }
 
-size_t Assembly::encodeSubPath(std::vector<size_t> const& _subPath)
+SubAssemblyID Assembly::encodeSubPath(std::vector<SubAssemblyID> const& _subPath)
 {
 	assertThrow(!_subPath.empty(), AssemblyException, "");
 	if (_subPath.size() == 1)
 	{
-		assertThrow(_subPath[0] < m_subs.size(), AssemblyException, "");
+		assertThrow(_subPath[0].value < m_subs.size(), AssemblyException, "");
 		return _subPath[0];
 	}
 
-	if (m_subPaths.find(_subPath) == m_subPaths.end())
+	if (!m_subPaths.contains(_subPath))
 	{
-		size_t objectId = std::numeric_limits<size_t>::max() - m_subPaths.size();
-		assertThrow(objectId >= m_subs.size(), AssemblyException, "");
+		SubAssemblyID const objectId{std::numeric_limits<SubAssemblyID::value_type>::max() - m_subPaths.size()};
+		assertThrow(objectId.value >= m_subs.size(), AssemblyException, "");
 		m_subPaths[_subPath] = objectId;
 	}
 
 	return m_subPaths[_subPath];
 }
 
-Assembly const* Assembly::subAssemblyById(size_t _subId) const
+Assembly const* Assembly::subAssemblyById(SubAssemblyID const _subId) const
 {
-	std::vector<size_t> subIds = decodeSubPath(_subId);
+	std::vector<SubAssemblyID> subIDs = decodeSubPath(_subId);
 	Assembly const* currentAssembly = this;
-	for (size_t currentSubId: subIds)
+	for (auto [subIDIndex]: subIDs)
```

Comment:
Wait, isn't this an ID rather than an index? You're iterating a `vector` of `SubAssemblyID`s. 
Though I also see that it gets cast to `size_t` without extracting the `.value`. Why does that even work?

This would be much clearer without `auto` obscuring the type.

```suggestion
	for (SubAssemblyID [subID]: subIDs)
```

Reply:
You also have some other loops using `size_t` `subIDIndex` where just using `SubAssemblyID` for the counter would make things simpler. E.g. in `optimiseInternal()` I think you could just do this:
```c++
for (SubAssemblyID subID = 0; subID.value < m_subs.size(); ++subID.value)
```

Reply:
It works because you can unpack structs much like tuples, `auto [subIDIndex] = subIDs.front()` already refers to the contained `value` member. Although `subIDIndex` was not a good name, should have been `subIDValue`.

In any case, I have changed the loop now to just iterate over the `SubAssemblyID` instances instead of contained members and then extract the index with a new `asIndex` method that does static casting and size checking if needed. That way there aren't that many static casts littered about.